### PR TITLE
DDP: coalescing many little broadcasts to improve performance

### PR DESCRIPTION
**Pull request by teng-li**
At the constructor, when we broadcast module 0 on Node 0's entire module states to all other nodes, we broadcast each state (tensor) one by one, this is very inefficient and also causes NCCL/Gloo deadlocks (previous, even though this is a different issue).

This PR coalesces the entire module states into a series of big Tensors and broadcasts them out in one shot, just like _sync_buffers() at each forward() function. 

The memory is limited by the bucket size, which is 10MB by default. 

This also reduces the chance that multi-process/node cross node NCCL deadlocks, which happens only after a long series of broadcast. So far after this change, I haven't seen any deadlocks yet.

Tested on ResNet50 training, I printed out the entire module's state on both nodes before the broadcast and after the broadcast, and verified that after the coalesced broadcast, node 0 and node 1 have the exact state and node 0's state hasn't changed either.  Thus, this PR should be safe to land

### CODE REVIEW

**Comment by Stonesjtu** [comment]
I'm curious how much time saving in this coalesced broadcast, and the overhead of packing/unpacking tensors. Should them be moved into a new contiguous memory.

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -188,6 +197,34 @@ def train(self, mode=True):
         for module in self._module_copies[1:]:
             module.train(mode)
 
+    def _dist_broadcast_coalesced(self, tensors, buffer_size):
```
Comment (teng-li) [inline torch/nn/parallel/distributed.py]: @apaszke  Now we bucket the coalescing with this function

**Comment by teng-li** [comment]
@Stonesjtu For about 100MB of broad across two nodes, I did a rough perf evaluation, with 2 singles broadcast vs the original many broadcast, we can nearly double the performance.  0.0268 sec (with 1 coalesced b-cast) vs 0.0436 sec (with the original logic).

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -100,9 +100,20 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
...
+        # used for intra-node param sync and inter-node sync as well
+        self.broadcast_bucket_size = 10 * MB
+
         # Sync params and buffers
-        for p in self.module.state_dict().values():
-            dist.broadcast(p, 0)
+        # Here we will coalesce all the parameters and buffers in several big
+        # flat tensors and broadcast them out to reduce the number of broadcasts
+        # as well as improve performance, even though this function is
+        # only called one time.
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: I don't think this comment is necessary. The important part is that we have to sync params. How we do it is clear from the code (these are just 2 lines).

File: torch/nn/parallel/distributed.py [hunk 3]
```diff
@@ -188,6 +197,34 @@ def train(self, mode=True):
...
+        """
+        tensors_bucket = []
+        # To init the first bucket right away
+        cur_bucket_size = buffer_size
+        for tensor in tensors:
+            if cur_bucket_size >= buffer_size:
+                tensors_bucket.append([])
+                cur_bucket_size = 0
+            tensors_bucket[-1].append(tensor)
+            cur_bucket_size += tensor.numel() * tensor.element_size()
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: There's `torch._utils._take_tensors` that will do most of these things for you, including handling different types, which is not handled here. Please use that. To see an example look at how `broadcast_coalesced` [used to be implemented](https://github.com/pytorch/pytorch/blob/fa5efab669d68f2931292176d3376e3519bdd0b4/torch/cuda/comm.py#L32-L70).

**Review by apaszke — COMMENTED** [review]
LGTM, but the code could be simpler and more robust if we used `_take_tensors`

**Review by apaszke — APPROVED** [review]

**Comment by ebetica** [comment]
@teng-li Is it possible to simply flatten the storage of all parameters (assuming they are static for the duration of the training)? As long as you do in place operations on parameters, which I _think_ PyTorch does, it seems like you'd be able to copy the entire model with just one broadcast.

**Comment by teng-li** [comment]
@ebetica hmm, yeah, good thought, but this enforces a static assumption, I am not too sure how feasible this would be. cc @apaszke

**Comment by apaszke** [comment]
We can't do this transformation in general, because there are cases where users want to flatten the parameters as they like, and we shouldn't interfere.

**Pull request merged**